### PR TITLE
Fixing TRUNK-6478 Adding hashCode to CohortMembership as well as upda…

### PR DESCRIPTION
**Pull request by fruether**
## Description of what I changed

- I added in CohortMembershipTest a new test case that checked if two identical CohortMemberships could be added to a set
- I added the equal signature of CohortMembership to use Object
- I added a hashCode method to CohortMembership

This is related to the following comment in a previous merge: https://github.com/openmrs/openmrs-core/commit/60d92bda69d0e46bda832e64bfd6b881320f911f#r31850302
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5478

## Checklist: I completed these to help reviewers :)

- [x ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/CohortMembership.java [hunk 1]
```diff
@@ -148,18 +149,29 @@ public int compareTo(CohortMembership o) {
...
-	 * 
-	 * @param otherCohortMembership is a CohortMembership object that should be checked for equality with this object
+	 *
+	 * @param otherCohortMembershipObject is a CohortMembership object that should be checked for equality with this object
 	 * @return true if both objects are logically equal. This is the case when endDate, startDate and patientId are equal  
 	 */
-	public boolean equals(CohortMembership otherCohortMembership) {
+	@Override
+	public boolean equals(Object otherCohortMembershipObject) {
+		if(!(otherCohortMembershipObject instanceof CohortMembership)){
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/CohortMembership.java]: Wouldn't it be cheaper for the `otherCohortMembership != null` check to be checked early here than at the `return` statement?

Comment (fruether) [reply api/src/main/java/org/openmrs/CohortMembership.java]: @samuelmale  Thank you for the review. I add the != null check to the top of the function! 

Do you have any further remarks

File: api/src/main/java/org/openmrs/CohortMembership.java [hunk 2]
```diff
@@ -51,7 +52,7 @@ public CohortMembership(Integer patientId) {
 	public boolean isActive(Date asOfDate) {
 		Date date = asOfDate == null ? new Date() : asOfDate;
 		return !this.getVoided() && OpenmrsUtil.compare(startDate, date) <= 0
-				&& OpenmrsUtil.compareWithNullAsLatest(date, endDate) <= 0;
+			&& OpenmrsUtil.compareWithNullAsLatest(date, endDate) <= 0;
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/CohortMembership.java]: Is this just a whitespace change?

Comment (fruether) [reply api/src/main/java/org/openmrs/CohortMembership.java]: Yes I think so. Not sure how this happened

Comment (samuelmale) [reply api/src/main/java/org/openmrs/CohortMembership.java]: Are able to revert most of the whitespace changes to improve our PR?

File: api/src/test/java/org/openmrs/CohortMembershipTest.java [hunk 3]
```diff
@@ -389,19 +392,52 @@ public void equal_shouldReturnTrueWhenBothPatientIdsAreNullAndEndDateAndStartDat
...
+	@Test
+	public void euqal_shouldPreventAddingTwoEqualObjectsToACollection() {
+		Set<CohortMembership> cohortMembershipSet = new LinkedHashSet<>();
+		
+		final int PATIENT_ID = 12;
+		CohortMembership cohortMembershipOne = new CohortMembership(PATIENT_ID);
+		CohortMembership cohortMembershipTwo = new CohortMembership(PATIENT_ID);
+		
+		final int TIMESTAMP_START_DATE = 1545140935;
+		final int TIMESTAMP_END_DATE = 1577836801;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/CohortMembershipTest.java]: If the constants are referenced in more than one testcase, why don't we make them class level constants?

File: api/src/main/java/org/openmrs/CohortMembership.java (hunk 2 above)
Comment (fruether) [reply api/src/main/java/org/openmrs/CohortMembership.java]: Done as requested. But I am not able to remove that space easily

**Comment by fruether** [comment]
@samuelmale i reverted the white space changes where easily applicable and I made the variables global. Any more issues you see that prevent these PR from merging

File: api/src/test/java/org/openmrs/CohortMembershipTest.java (hunk 3 above)
Comment (fruether) [reply api/src/test/java/org/openmrs/CohortMembershipTest.java]: Done

**Pull request merged**